### DISCUSSION TITLE
'with' assigns a value to the index

### DISCUSSION THREAD

**Discussion by ilyfairy** [dotnet/csharplang#6765] [Language Ideas]
'with' assigns a value to the index
```CSharp
int[] arr = { 1, 2, 3 };
int[] newArr = arr with { [0] = arr[^1], [^1] = arr[0] };
// is equal to the following code
int[] newArr = arr[..];
(newArr[0],newArr[0]) = (arr[^1],arr[0]);

// and the following
var span = stackalloc int[10] with { [0] = 1, [1] = 2 };
```

**Comment by jnm2** [comment]
An alternative could potentially be the syntax `[..arr, 0: 1, 1: 2]`. https://github.com/dotnet/csharplang/issues/5354 plans to bring the syntax `[..dict, key1: newValue1, key2: newValue2]` to create a cloned dictionary with the specified updates. It was an unresolved question whether this would only bind to an indexer mapped to `IDictionary<TKey, TValue>.this[TKey]` or any accessible indexer on the concrete collection type. The latter would allow this to work for non-dictionary collections such as arrays. @RikkiGibson had some thoughts about custom collection types.

**Comment by jnm2** [comment]
The suggested syntax complements https://github.com/dotnet/csharplang/discussions/4889.

What if this could combine with complex nested `with` expressions (https://github.com/dotnet/csharplang/discussions/77) to allow `with { A[0].B = c }` and other constructions?
